### PR TITLE
Feature: endpoints

### PR DESCRIPTION
**Pull request by vadymmarkov**
This is a super breaking change, so the next version should be **Malibu 3**.

We used structs to describe requests  for a while and it worked, but it wasn't the most elegant solution. I see libraries like https://github.com/Moya/Moya and https://github.com/kickstarter/ios-ksapi/blob/master/KsApi/lib/Route.swift use `enum` for this and I think it looks really nice. Based on this idea I've made the following API changes:

**1. Endpoint**

Additional headers, base url and request description are moved to `Endpoint`. There are no `Requestable` protocols anymore, they are replaced by `Request` struct with helper `get`, `post`, `patch`, `put`, `delete` and `head` methods:

```swift
enum SharkywatersService: Endpoint {
  // Describe requests
  case fetchBoards
  case showBoard(id: Int)
  case createBoard(type: Int, title: String)
  case updateBoard(id: Int, type: Int, title: String)
  case deleteBoard(id: Int)

  // Every request will be scoped by the base url
  static var baseUrl: URLStringConvertible = "http://sharkywaters.com/api/"

  // Additional headers for every request
  static var headers: [String: String] = [
    "Accept" : "application/json"
  ]

  // Build requests
  var request: Request {
    switch self {
    case .fetchBoards:
      return Request.get("boards")
    case .showBoard(let id):
      // Let's use JSON dictionary as a mock data
      return Request.get("boards:\(id)", mock: Mock(json: ["type": 1, "title": "Classic"]))
    case .createBoard(let type, let title):
      return Request.post("boards", parameters: ["type": type, "title": title])
    case .updateBoard(let id, let title):
      return Request.patch("boards\(id)", parameters: ["title": title])
    case .deleteBoard(let id):
      return Request.delete("boards\(id)")
    }
  }
}

let networking = Networking<SharkywatersService>()
networking.request(.fetchBoards)
```

**2. Networking is endpoint-specific:**

```swift
let networking = Networking<SharkywatersService>(
  // `OperationQueue` Mode
  mode: .async,
  // Mock behavior (never, delayed)
  mockBehavior: .delayed(0.5),
  // `default`, `ephemeral`, `background` or `custom`
  sessionConfiguration: .default,
  // Custom `URLSessionDelegate` could set if needed
  sessionDelegate: self
)
```

**3. There is no networking container anymore, it's up to developer to create and keep `Networking` instances.**

**4. Mocks are created directly on a request:**

```swift
let request = Request.get(
  "boards",
  mock: Mock(fileName: "boards.json")
)
```

I think it simplifies the way requests are described and used, plus it's gonna remove a lot of boilerplate code from our projects.

### CODE REVIEW

File: MalibuTests/Helpers/Mocks.swift [hunk 1]
```diff
@@ -2,49 +2,40 @@ import Foundation
...
+  var request: Request {
+    switch self {
+    case .fetchPosts:
+      return Request.get("posts")
+    case .showPost(let id):
+      return Request.get("posts:\(id)", mock: Mock(json: ["title": "Test"]))
+    case .createPost(let title):
+      return Request.post("posts", parameters: ["title": title])
+    case .replacePost(let id, let title):
+      return Request.put("posts\(id)", parameters: ["title": title])
```
Comment (onmyway133) [inline MalibuTests/Helpers/Mocks.swift]: @vadymmarkov is this typo, should it be `posts/\(id)`?

File: MalibuTests/Helpers/Mocks.swift [hunk 2]
```diff
@@ -2,49 +2,40 @@ import Foundation
...
+
+  static var baseUrl: URLStringConvertible = "http://api.loc"
+  static var headers: [String: String] = [:]
+
+  var request: Request {
+    switch self {
+    case .fetchPosts:
+      return Request.get("posts")
+    case .showPost(let id):
+      return Request.get("posts:\(id)", mock: Mock(json: ["title": "Test"]))
```
Comment (onmyway133) [inline MalibuTests/Helpers/Mocks.swift]: What is `:`?

File: MalibuTests/Helpers/Mocks.swift (hunk 1 above)
Comment (vadymmarkov) [inline MalibuTests/Helpers/Mocks.swift]: Yes, it is a typo.

**Comment by onmyway133** [comment]
Looks great 🚀

**Comment by onmyway133** [comment]
@vadymmarkov Look like travis is yelling "TEST FAIL"

**Comment by vadymmarkov** [comment]
@onmyway133 I can't understand what is wrong there. Tests are passing on my machine 😬

**Comment by onmyway133** [comment]
💥

**Comment by viktorgardart** [comment]
🎉 🍾

**Comment by zenangst** [comment]
This looks awesome!

**Pull request merged**